episodes/scraping-data.md

```diff
@@ -6,24 +6,581 @@ exercises: 0
+```
+
+Oh no! We get an error:
+
+```output
+MissingSchema: Invalid URL 'xls/f923_2024.zip': No scheme supplied. Perhaps you meant https://xls/f923_2024.zip?
+```
+
+Looks like the URL in the `href` is incomplete. It turns out that this is a *relative path* - much like the relative paths you had to deal with when loading data on your computer. The full URL we want is
+`https://www.eia.gov/electricity/data/eia923/xls/f923_2024.zip` - which combines the URL of the page we got the link from (`https://www.eia.gov/electricity/data/eia923/eia906u.php`) with the fragment we got in the `href` (`xls/f923_2024.zip`).
+
+This is a super common thing to have to do, so there's a useful bit of the Python standard library for this: `urllib.parse.urljoin`:
+
+```
+from urllib.parse import urljoin
+
+urljoin(eia_923_url, "xls/f923_2024.zip")
+```
+
+We can also do the string concatenation ourselves with `eia_906_url + "..."`, but there are a surprising amount of details to get wrong here so it's nice to just use the function that works.
+
+Let's add that into our loop:
+
+```
+for a in eia_923_links:
+  full_link = urljoin(eia_923_url, a["href"])
+  response = requests.get(a["href"])
+  eia_923_blobs.append(response.content)
+```
+:::: challenge
+
+#### Challenge: get the Form 906 file contents
+
+OK, so now we know how to scrape a bunch of URLs from a webpage. Let's read the Form 906 files into our program! Since they're XLS files, we can read them directly from a URL using `pandas.read_excel` - no need for manually downloading with `requests`.
+
+Try filling out the body of this for loop.
+
+```python
+from urllib.parse import urljoin
+
+import bs4
+import requests
+import pandas as pd
+
+eia_906_url = "https://www.eia.gov/electricity/data/eia923/eia906u.php"
+response = requests.get(eia_906_url)
+soup = bs4.BeautifulSoup(response.text)
+
+a_tags = soup.find_all("a", href=True)
+eia_906_links = []
+for a in a_tags:
+    if ".xls" in a["href"].lower():
+        eia_906_links.append(a)
+
+eia_906_dataframes = []
+for a in eia_906_links:
+    ...
+```
+
+::::::: solution
+
+```python
+for a in eia_906_links:
+    full_url = urljoin(eia_906_url, a["href"])
+    eia_906_dataframes = pandas.read_excel(full_url)
+```
+:::::::
+
+::::
+
+Once we've completed the challenge above, we have a list of a bunch of dataframes. To bring them all into one dataframe, we can use `pd.concat`:
+
+```python
+mega_906 = pd.concat(eia_906_dataframes)
+```
+
+If you use `DataFrame.info()` you can quickly see that some columns (YEAR, FIPST, UTILNAME) are more populated than others (MULTIST, GEN01, etc):
+
+```python
+mega_906.info()
+```
+
+And if you start to dig into the data a bit, such as pulling out the various values of `YEAR`, you see that you have *plenty* of data cleaning to do before this is really usable for analysis. But at least you have all of the data in one dataframe now! We'll go over some more tips for exploratory data analysis in the next episode.
```

Comment:
Not sure if this is a useful tangent or not... is it worth poking around the resulting dataframe a little and maybe motivating your EDA stuff in lesson 2.1?